starport/pkg/placeholder/tracer.go

```diff
@@ -0,0 +1,129 @@
+package placeholder
+
+import (
+	"strings"
+
+	"github.com/tendermint/starport/starport/pkg/validation"
+)
+
+type iterableStringSet map[string]struct{}
+
+func (set iterableStringSet) Iterate(f func(i int, element string) bool) {
+	i := 0
+	for key := range set {
+		if !f(i, key) {
+			return
+		}
+		i++
+	}
+}
+
+func (set iterableStringSet) Add(item string) {
+	set[item] = struct{}{}
+}
+
+var _ validation.Error = (*ErrMissingPlaceholders)(nil)
+
+type ErrMissingPlaceholders struct {
+	missing        iterableStringSet
+	additionalInfo string
+}
+
+// Is true if both errors have the same list of missing placeholders.
+func (e *ErrMissingPlaceholders) Is(err error) bool {
+	other, ok := err.(*ErrMissingPlaceholders)
+	if !ok {
+		return false
+	}
+	if len(other.missing) != len(e.missing) {
+		return false
+	}
+	for i := range e.missing {
+		if e.missing[i] != other.missing[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func (e *ErrMissingPlaceholders) Error() string {
+	var b strings.Builder
+	b.WriteString("missing placeholders: ")
+	e.missing.Iterate(func(i int, element string) bool {
+		if i > 0 {
+			b.WriteString(", ")
+		}
+		b.WriteString(element)
+		return true
+	})
+	return b.String()
+}
+
+func (e *ErrMissingPlaceholders) ValidationInfo() string {
+	var b strings.Builder
+	b.WriteString("Missing placeholders:\n\n")
+	e.missing.Iterate(func(i int, element string) bool {
+		if i > 0 {
+			b.WriteString("\n")
+		}
+		b.WriteString(element)
+		return true
+	})
+	if e.additionalInfo != "" {
+		b.WriteString("\n\n")
+		b.WriteString(e.additionalInfo)
+	}
+	return b.String()
+}
+
+// Option for configuring session.
+type Option func(*Tracer)
+
+// WithAdditionalInfo will append info to the validation error.
+func WithAdditionalInfo(info string) Option {
+	return func(s *Tracer) {
+		s.additionalInfo = info
+	}
+}
+
+// New instantiates Session with provided options.
+func New(opts ...Option) *Tracer {
+	s := &Tracer{missing: iterableStringSet{}}
+	for _, opt := range opts {
+		opt(s)
+	}
+	return s
+}
+
+type Replacer interface {
+	Replace(content, placeholder, replacement string) string
+}
+
+// Tracer keeps track of missing placeholders.
+type Tracer struct {
+	missing        iterableStringSet
```

Comment:
Not a blocker but why not just using a string slice without the `Add` and `Iterate` funcs to make the code simpler?

Reply:
some strings are used more then once